### PR TITLE
[14.0][FIX] purchase_sale_inter_company: sync confirmed unlocked purchase updates

### PR DESCRIPTION
**Pull request by chienandalu**
When the purchase user updates the PO after the SO is generated and always if both are confirmed and unlocked, we sync the new lines to update the demand.

cc @Tecnativa TT46773

please take a look @pedrobaeza

### CODE REVIEW

File: purchase_sale_inter_company/models/purchase_order.py [hunk 1]
```diff
@@ -186,3 +186,38 @@ def button_cancel(self):
...
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        """Sync lines between an open purchase and an open sale order"""
+        lines = super().create(vals_list)
+        for order in lines.order_id.filtered(lambda x: x.state == "purchase"):
+            sale = self.env["sale.order"].search(
+                [("auto_purchase_order_id", "=", order.id)]
```
Comment (pedrobaeza) [inline purchase_sale_inter_company/models/purchase_order.py]: Why not adding a o2m for avoiding this ugly syntax?

**Review by pedrobaeza — COMMENTED** [review]
I miss a `write` override for also synchronizing changed quantities/prices.

File: purchase_sale_inter_company/models/purchase_order.py (hunk 1 above)
Comment (chienandalu) [reply purchase_sale_inter_company/models/purchase_order.py]: It looks like o2m relations are limited by the record rules and I couldn't access it even with `sudo()`. An alternative approach has been a compute field with `compute_sudo`

**Comment by chienandalu** [comment]
> I miss a write override for also synchronizing changed quantities/prices.

I added that logic as well. For the moment only to sync quantities

**Comment by francesco-ooops** [comment]
> When the purchase user updates the PO after the SO is generated and always if both are open, we sync the new lines to update the demand.

thanks for the PR @chienandalu , just to understand better functionally: what states are you referring to with "open"?

**Comment by chienandalu** [comment]
> thanks for the PR @chienandalu , just to understand better functionally: what states are you referring to with "open"?

Basically, both the purchase and its sale order are in states `purchase/sale`

**Comment by francesco-ooops** [comment]
@chienandalu ok, so they're both "confirmed" :)

Will test and review

**Comment by chienandalu** [comment]
> @chienandalu ok, so they're both "confirmed" :)

Yes, but not `done` (locked)

File: purchase_sale_inter_company/models/purchase_order.py [hunk 2]
```diff
@@ -3,13 +3,30 @@
...
 
-from odoo import _, models
+from odoo import _, api, fields, models
 from odoo.exceptions import UserError
 
 
 class PurchaseOrder(models.Model):
     _inherit = "purchase.order"
 
+    auto_sale_order_id = fields.Many2one(
```
Comment (pedrobaeza) [inline purchase_sale_inter_company/models/purchase_order.py]: Let's name properly new fields (and the same for the other):

```suggestion
    intercompany_sale_order_id = fields.Many2one(
```

File: purchase_sale_inter_company/models/purchase_order.py [hunk 3]
```diff
@@ -186,3 +203,107 @@ def button_cancel(self):
...
+            self.env["sale.order.line"].with_user(intercompany_user.id).sudo().create(
+                sale_lines
+            )
+        return lines
+
+    @api.model
+    def _get_purchase_sale_line_sync_fields(self):
+        """Map purchase line fields to the synced sale line peer"""
+        return {
+            "product_qty": "product_uom_qty",
```
Comment (pedrobaeza) [inline purchase_sale_inter_company/models/purchase_order.py:265]: Please include price_unit here (and discount should be added in a glue module).

File: purchase_sale_inter_company/models/stock_picking.py [hunk 4]
```diff
@@ -30,13 +30,28 @@ def _action_done(self):
...
-                    raise UserError(
-                        _(
-                            "Mismatch between move lines with the "
-                            "corresponding  PO %s for assigning "
-                            "quantities and lots from %s for product %s"
-                        )
-                        % (purchase.name, pick.name, move.product_id.name)
+                    note = _(
+                        "Mismatch between move lines with the "
+                        "corresponding  PO %s for assigning "
```
Comment (pedrobaeza) [inline purchase_sale_inter_company/models/stock_picking.py]:
```suggestion
                        "corresponding PO %s for assigning "
```

File: purchase_sale_inter_company/models/purchase_order.py (hunk 3 above)
Comment (chienandalu) [reply purchase_sale_inter_company/models/purchase_order.py:265]: IMO the price of the sale order is the one that should prevail. We don't really want to sync that value from the PO to the SO (although we could want to do it back but we don't have the mechanism at this moment)

Comment (pedrobaeza) [reply purchase_sale_inter_company/models/purchase_order.py:265]: OK, please add it to the roadmap.

File: purchase_sale_inter_company/models/stock_picking.py (hunk 4 above)
Comment (chienandalu) [reply purchase_sale_inter_company/models/stock_picking.py]: done

File: purchase_sale_inter_company/models/purchase_order.py (hunk 3 above)
Comment (chienandalu) [reply purchase_sale_inter_company/models/purchase_order.py:265]: done

File: purchase_sale_inter_company/models/purchase_order.py (hunk 2 above)
Comment (chienandalu) [reply purchase_sale_inter_company/models/purchase_order.py]: done

**Review by pedrobaeza — APPROVED** [review]

**Review by francesco-ooops — APPROVED** [review]
Sync works well with following use case: 

Chicago purchases from San Francisco, product Acoustic Block Screen qty 5 > confirm
Chicago updates POL qty to 10 
San Francisco finds qty updated to 10 on SOL and picking OUT

It doesn't work well with flow: 

Chicago purchases from San Francisco, product Acoustic Block Screen qty 10 > confirm
Chicago updates POL qty to 5
San Francisco still has qty 10 on SOL and picking OUT

Seems standard behavior for PO tho... I think at least there should be a warning like there is in SO, "You are decreasing the ordered quantity! Do not forget to manually update the delivery order if needed."

**Comment by chienandalu** [comment]
> Chicago purchases from San Francisco, product Acoustic Block Screen qty 10 > confirm
> Chicago updates POL qty to 5
> San Francisco still has qty 10 on SOL and picking OUT

I just tested it on runboat and it works as expected:

- the sale order gets the updated qty to the decreased one and the picking schedules an warning activity for the decreased qty (this is the standard in this version when you decrease the demand from the sale order)

**Comment by pedrobaeza** [comment]
Please rebase, as there's now a conflict.

**Comment by chienandalu** [comment]
- Rebased
- Added a test with @francesco-ooops case

**Comment by pedrobaeza** [comment]
Please fw-port it.

**Pull request merged**